### PR TITLE
onboardingTicket: add uid to the ticket

### PR DESCRIPTION
**Pull request by rewantsoni**
onboarding ticket should contain the storageCluster UID the token was generated for and to fetch the storageCluster for UID, we fetch the storage Cluster from operator namespace

### CODE REVIEW

**Comment by mrudraia1** [comment]
I faced storagecluster UID error while testing the PR 2827. The PR 2827 can be tested after this PR is merged.

File: services/ux-backend/handlers/onboarding/peertokens/handler.go
Comment (nb-ohad) [inline services/ux-backend/handlers/onboarding/peertokens/handler.go:37]: Why so we need the namespaced name? Don't we just need a namespace?

File: services/ux-backend/handlers/onboarding/peertokens/handler.go
Comment (nb-ohad) [inline services/ux-backend/handlers/onboarding/peertokens/handler.go:70]: Can't we accept UID as the parameter type for the function? It makes more sense that we would like to keep the param as typed as possible

File: services/ux-backend/handlers/onboarding/peertokens/handler.go
Comment (nb-ohad) [inline services/ux-backend/handlers/onboarding/peertokens/handler.go:68]: This entire logic here needs to be duplicated to the client onboarding token handler. I would suggest to have this as a util of sorts to be invoked from both places

File: services/ux-backend/main.go
Comment (nb-ohad) [inline services/ux-backend/main.go:129]: 1. I believe we already have this code in the code base somewhere else. 
I would prefer to have a shared code invoked from all locations. 

2. In this file all private function resides above the main function, can we please be consistency

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: services/ux-backend/handlers/onboarding/peertokens/handler.go
Comment (rewantsoni) [reply services/ux-backend/handlers/onboarding/peertokens/handler.go:37]: I was thinking from a user perspective, it would make more sense to specify the storageCluster Name they want to onboard to instead of the namespace of the storageCluster. Since storageCluster is a namespace resource, we can have storageClusters with the same name in a different namespace hence chose to go with storageCluster namespacedName to identify the storageCluster user wants to onboard to

Comment (nb-ohad) [reply services/ux-backend/handlers/onboarding/peertokens/handler.go:35]: These endpoints are mainly for the UI to invoke, the UI already has the namespace context and does not need to ask the user for anything. Also, a namespace name as part of a user ask can create mistaken input, where the named resource does not exist in the namespace.

File: controllers/util/client.go
Comment (leelavg) [inline controllers/util/client.go:1]: since we have atleast two clients, gRPC and Kube better rename this file to be more accurate? `k8sclient.go`?

File: controllers/util/k8sutil.go
Comment (leelavg) [inline controllers/util/k8sutil.go:170]: function name already says `Namespace` and last arg is repeating the same, remove a duplicate from one of these? `GetStorageCluster` or `namespace string`?

File: onboarding-validation-keys-generator/main.go
Comment (leelavg) [inline onboarding-validation-keys-generator/main.go:60]: can be replaced with newly created function `util.GetStorageCluster`?

File: services/ux-backend/handlers/onboarding/peertokens/handler.go
Comment (leelavg) [inline services/ux-backend/handlers/onboarding/peertokens/handler.go:33]: could you pls follow the existing way of error handling across the service?

File: services/ux-backend/main.go
Comment (leelavg) [inline services/ux-backend/main.go:70]: we are starting w/ supporting a single storagecluster that exists in operator ns?

File: services/ux-backend/main.go
Comment (leelavg) [inline services/ux-backend/main.go:74]:
```suggestion
		klog.Exitf("failed to add corev1 to scheme. %v", err)
```

File: services/ux-backend/main.go
Comment (leelavg) [inline services/ux-backend/main.go:82]:
```suggestion
		klog.Exitf("failed to create kube client: %v", err)
```

File: onboarding-validation-keys-generator/main.go
Comment (leelavg) [inline onboarding-validation-keys-generator/main.go:31]:
```suggestion
		klog.Exitf("failed to add corev1 to scheme. %v", err)
```

File: services/ux-backend/main.go
Comment (rewantsoni) [reply services/ux-backend/main.go:70]: Yes

File: controllers/util/k8sutil.go
Comment (leelavg) [reply controllers/util/k8sutil.go]: thinking again, suggestions may just show up function name and in that scenario as we are using in multiple places, `GetStorageClusterByNamespace` would be good?

File: controllers/util/k8sutil.go
Comment (nb-ohad) [inline controllers/util/k8sutil.go:172]: Can you please cap this list operation to 1 result?

File: controllers/util/provider.go
Comment (nb-ohad) [inline controllers/util/provider.go:55]: 1. Please cap only the first letter of an abbreviation 
2. Why do we need to call this StorageClusterUID and not just StorageCluster in the context of the ticket?

File: controllers/util/provider.go
Comment (nb-ohad) [inline controllers/util/provider.go:46]: Please cap only the first letter of an abbreviation 
```suggestion
func GeneratePeerOnboardingToken(tokenLifetimeInHours int, privateKeyPath string, storageClusterUid types.UID) (string, error) {
```

File: services/types.go
Comment (nb-ohad) [inline services/types.go:17]: Only capital the first latter of an abbreviations
```suggestion
	StorageClusterUid types.UID             `json:"storageClusterUid"`
```

File: services/ux-backend/handlers/onboarding/peertokens/handler.go
Comment (nb-ohad) [reply services/ux-backend/handlers/onboarding/peertokens/handler.go:33]: @rewantsoni You are in the context of handling the message, please follow the proper way to write into the response stream (w.WriteHeader), and use a proper HTTP response code

File: controllers/util/k8sclient.go
Comment (nb-ohad) [inline controllers/util/k8sclient.go:10]: 1. There is no need to have a new util file. This is definitely a k8s util function.
2. Moving this into util is a good move, but you still need to consolidate all the calla (for client creation) across the repo

**Review by nb-ohad — CHANGES_REQUESTED** [review]

**Review by nb-ohad — COMMENTED** [review]
@rewantsoni I see the changes to the token done for the `peertokens` handler, but where are the changes to the `peerclient` handler? 
The introduction of a storagecluster UID into the token need to happen there as well (it should be a mandatory field)

**Comment by rewantsoni** [comment]
>@rewantsoni You are in the context of handling the message, please follow the proper way to write into the response stream (w.WriteHeader), and use a proper HTTP response code
@nb-ohad http.Error is a helper function that sets the error code on the header and sets the response. See [here](https://pkg.go.dev/net/http#Error)

**Comment by rewantsoni** [comment]
>@rewantsoni You are in the context of handling the message, please follow the proper way to write into the response stream (w.WriteHeader), and use a proper HTTP response code
@nb-ohad http.Error is a helper function that sets the error code on the header and sets the response. See [here](https://pkg.go.dev/net/http#Error)

File: controllers/util/k8sutil.go
Comment (rewantsoni) [reply controllers/util/k8sutil.go:175]: I was going through https://github.com/red-hat-storage/ocs-operator/pull/2837, and from what I understood we can have multiple storageClusters, but the other storageCluster's phase will be ignored. Modified the function there to GetStorageClusterInNamespace

File: controllers/util/provider.go
Comment (rewantsoni) [reply controllers/util/provider.go]: golang-lint is failing with this change

File: controllers/util/k8sutil.go
Comment (nb-ohad) [reply controllers/util/k8sutil.go:175]: You cannot have multi-storage clusters in a single namespace, and this lists only inside the namespace. So my ask adding `client.Count(1)` is meaningful

Comment (malayparida2000) [reply controllers/util/k8sutil.go:175]: We can have multple storageclusters inside a single namespace, Just that except one all else would be in phase ignored. Limiting the list to 1 result can easily give us the wrong result with a phasIgnored storageCluster. I would suggest listing all the items & then filtering from that list based on the requirement is the better approach.

Comment (nb-ohad) [reply controllers/util/k8sutil.go:175]: @malayparida2000 I understand what you are saying, but that is a scenario we should not address this way. That is a user error, and if we want to solve it we should introduce a webhook to prevent the creation of a second storage cluster. 

Listing with a count of (1) prevents the deep copy operations of all other storage classes from the cache into the resulting list, in the case of storage cluster it is a very very big structure (even if it is mostly not filled) and we should avoid the deep copy.

File: controllers/util/provider.go
Comment (nb-ohad) [reply controllers/util/provider.go:47]: I think it is a golang lint issue, @iamniting might be able to give some advice here, or suggest a separate fix for this.
In the meantime, we can merge this (assuming no other issues) without this change.

File: services/ux-backend/handlers/onboarding/clienttokens/handler.go
Comment (nb-ohad) [inline services/ux-backend/handlers/onboarding/clienttokens/handler.go:67]: This cannot be an `http.StatusBadRequest` because the failure has nothing to do with the content of the request, which also means that the user cannot change the request and clear the check. 
For these kinds of issues, I would suggest using `http.StatusInternalServerError`

File: services/ux-backend/handlers/onboarding/peertokens/handler.go
Comment (nb-ohad) [inline services/ux-backend/handlers/onboarding/peertokens/handler.go:33]: Same as my last comment. This cannot be a `http.StatusBadRequest`

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/util/k8sutil.go
Comment (rewantsoni) [reply controllers/util/k8sutil.go:175]: @malayparida2000 @nb-ohad I will open an issue about it, so that we can create a webhook for it after the feature freeze, for now, we can list with a count of (1) and ignore the others

Comment (rewantsoni) [reply controllers/util/k8sutil.go:175]: Issue: https://github.com/red-hat-storage/ocs-operator/issues/2877

Comment (leelavg) [reply controllers/util/k8sutil.go]: pls excuse for pitching in and correct me if I'm not understanding it properly. Agreed, it's an user error but if it did happen, count of (1) has a min of 50% chance of going wrong.

until we make the storagecluster in a ns singleton I don't think we incur a lot of mem usage if runtime does a deep copy to provide us as the stack size can go upto MBs and last time I checked storagecluster is around 2KB (without managed fields) but w/ pointer and all it'll more than that but not in a 2 digit MBs.

however, I agree w/ Ohad since we are returning a pointer to caller and storagecluster escapes to heap so does the list that is queried in this function increasing GC and heap usage, but kube clients outside of manager aren't backed up by cache and so every list call hits the api server (with the limit if set).

imo, we should code for correctness and list all storageclusters to filter out ignored one's unless we make storagecluster a singleton before merging this PR.

Comment (malayparida2000) [reply controllers/util/k8sutil.go]: I completely agree with Leela. Whether or not we decide to enforce having a single StorageCluster as mandatory is a separate discussion. However, currently, limiting the results to just one could lead to incorrect outcomes if we actually have more than one StorageCluster. So we can not limit the result to count(1) as of now.

Regarding the idea of creating a webhook to prevent additional StorageCluster creation, let's consider the practical implications. In most cases, a customer would either have a single StorageCluster as intended, or they might mistakenly create a second one. It's unlikely that a customer would intentionally create hundreds of ignored StorageCluster CRs to overwhelm our cache. Given this, the overhead of implementing a webhook to restrict additional StorageCluster creation might outweigh listine an extra StorageCluster in the "ignored" phase if one exists due to a customer error.

Comment (nb-ohad) [reply controllers/util/k8sutil.go]: @leelavg The point is that the probability of that happening is very low and we will inform that on the log and will instruct the user/support to delete the other StorageCluster. The code as it is now just ignores the additional storage cluster which is the wrong thing to do! It should be an error. 

now, considering the amount of places and flows where this method is invoked and the count of deep copies this method will create for a very big struct (storage cluster). I am asking us to treat an additional storage cluster as an error not as thing that is just ignored

@rewantsoni please make that change as requested

File: controllers/util/k8sutil.go
Comment (nb-ohad) [inline controllers/util/k8sutil.go:184]: Coming from the discussion we had on the other comment thread. You need to check that the storage cluster that you fetched is not in ignored state before proceeding. And if it is then you should return an error saying that this SC need to be deleted.

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/util/k8sutil.go
Comment (nb-ohad) [reply controllers/util/k8sutil.go]: @malayparida2000 
> Given this, the overhead of implementing a webhook to restrict additional StorageCluster creation might outweigh listine an extra StorageCluster in the "ignored" phase if one exists due to a customer error.

We can deliberate about the nececity for a webhook at a later time, that has nothing to do with the fact that the system is misconfigured and we just ignoring it.

**Pull request merged**
